mobius-core/src/main/java/com/spotify/mobius/ConditionalFilteringEventSource.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * -\-\-
+ * Mobius
+ * --
+ * Copyright (c) 2017-2020 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius;
+
+import static com.spotify.mobius.internal_util.Preconditions.checkNotNull;
+
+import com.spotify.mobius.disposables.Disposable;
+import com.spotify.mobius.functions.Consumer;
+import com.spotify.mobius.functions.Function;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+/**
+ * Allows the creation of {@link FilteringEventSource} based on a generic {@link EventSource} and a
+ * condition determiner function.
+ *
+ * @param <T> The Event Class to be emitted, and which is used in determining how to filter other
+ *     event sources.
+ */
+public final class ConditionalFilteringEventSource<T> implements FilteringEventSource<T> {
+
+  @Nonnull private final EventSource<T> source;
+  @Nonnull private final Function<T, Boolean> condition;
+
+  @Nonnull
+  final List<WeakReference<FilteredEventSource<?>>> filteredEventSources = new ArrayList<>(1);
```

Comment:
This should probably also be a CopyOnWriteArrayList?

Reply:
Yeah, this one should be fine to convert.